packages/server-core/src/user/identity-provider/identity-provider.hooks.ts

```diff
@@ -44,56 +52,138 @@ import {
+  // since the guest user will be destroyed once they log in
+  if (thisIdentityProvider.type === 'guest') return context
+
+  const providers = await context.app
+    .service(identityProviderPath)
+    .find({ query: { userId: thisIdentityProvider.userId } })
+
+  if (providers.total <= 1) {
+    throw new MethodNotAllowed('Cannot remove the only identity provider on a user')
+  }
+  return context
+}
+
+function addUserId(key: 'data' | 'query') {
+  return (context: HookContext) => {
+    if (key === 'data') {
+      context.data.userId = context.existingUser!.id
     } else {
-      // If trying to CRUD multiple identity-providers, e.g. patch all IP's belonging to a user, make params.query.userId
-      // the ID of the calling user, so no one can alter anyone else's IPs.
-      const userId = context.params[identityProviderPath]?.userId
-      if (!userId) throw new NotFound()
-      if (!context.params.query) context.params.query = {}
-      context.params.query.userId = userId
+      if (context.params.provider) context.params.query.userId = context.params.user!.id
     }
-    if (context.data) context.data = { password: context.data.password } //If patching externally, should only be able to change password
-    return context
   }
 }
 
-const checkOnlyIdentityProvider = () => {
-  return async (context: HookContext): Promise<HookContext> => {
-    if (!context.id) {
-      // do not allow to remove identity providers in bulk
-      throw new MethodNotAllowed('Cannot remove multiple providers together')
-    }
-    const thisIdentityProvider = (await context.app
-      .service(identityProviderPath)
-      .get(context.id)) as IdentityProviderType
+/** (BEFORE) CREATE HOOKS **/
 
-    if (!thisIdentityProvider) throw new Forbidden('You do not have any identity provider')
+// async function validateAuthParams(context:HookContext<IdentityProviderService>) {
+async function validateAuthParams(context: HookContext) {
+  let userId = context.data.userId
 
-    // we only want to disallow removing the last identity provider if it is not a guest
-    // since the guest user will be destroyed once they log in
-    if (thisIdentityProvider.type === 'guest') return context
+  if (context.params.authentication) {
+    const authResult = await (context.app.service('authentication') as any).strategies.jwt.authenticate(
+      { accessToken: context.params.authentication.accessToken },
+      {}
+    )
+    userId = userId || authResult[appConfig.authentication.entity]?.userId
+  }
 
-    const providers = await context.app
-      .service(identityProviderPath)
-      .find({ query: { userId: thisIdentityProvider.userId } })
+  if (!userId) {
+    return
+  }
+
+  const user = await context.app.service(userPath).get(userId)
+
+  context.existingUser = user
+}
 
-    if (providers.total <= 1) {
-      throw new MethodNotAllowed('Cannot remove the only identity provider on a user')
+async function addIdentityProviderType(context: HookContext) {
```

Comment:
All hooks should be typcasted as `HookContext<IdentityProviderPath>`. HookContext can be imported from declarations.ts